// @theia/core/lib/common/message-service.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var message_service_protocol_1 = require("./message-service-protocol");
var MessageService = /** @class */ (function () {
    function MessageService(client) {
        this.client = client;
    }
    // tslint:disable-next-line:no-any
    MessageService.prototype.log = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.processMessage(message_service_protocol_1.MessageType.Log, message, args);
    };
    // tslint:disable-next-line:no-any
    MessageService.prototype.info = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.processMessage(message_service_protocol_1.MessageType.Info, message, args);
    };
    // tslint:disable-next-line:no-any
    MessageService.prototype.warn = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.processMessage(message_service_protocol_1.MessageType.Warning, message, args);
    };
    // tslint:disable-next-line:no-any
    MessageService.prototype.error = function (message) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.processMessage(message_service_protocol_1.MessageType.Error, message, args);
    };
    // tslint:disable-next-line:no-any
    MessageService.prototype.processMessage = function (type, text, args) {
        if (!!args && args.length > 0) {
            var first = args[0];
            var actions = args.filter(function (a) { return typeof a === 'string'; });
            var options = (typeof first === 'object' && !Array.isArray(first))
                ? first
                : undefined;
            return this.client.showMessage({ type: type, options: options, text: text, actions: actions });
        }
        return this.client.showMessage({ type: type, text: text });
    };
    MessageService = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(message_service_protocol_1.MessageClient)),
        __metadata("design:paramtypes", [message_service_protocol_1.MessageClient])
    ], MessageService);
    return MessageService;
}());
exports.MessageService = MessageService;
