// @theia/workspace/lib/browser/workspace-commands.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var uri_1 = require("@theia/core/lib/common/uri");
var diff_uris_1 = require("@theia/core/lib/browser/diff-uris");
var selection_service_1 = require("@theia/core/lib/common/selection-service");
var common_frontend_contribution_1 = require("@theia/core/lib/browser/common-frontend-contribution");
var filesystem_1 = require("@theia/filesystem/lib/common/filesystem");
var dialogs_1 = require("@theia/core/lib/browser/dialogs");
var browser_1 = require("@theia/core/lib/browser");
var uri_command_handler_1 = require("@theia/core/lib/common/uri-command-handler");
var workspace_service_1 = require("./workspace-service");
var message_service_1 = require("@theia/core/lib/common/message-service");
var validFilename = require('valid-filename');
var WorkspaceCommands;
(function (WorkspaceCommands) {
    WorkspaceCommands.OPEN = {
        id: 'workspace:open',
        label: 'Open...'
    };
    WorkspaceCommands.CLOSE = {
        id: 'workspace:close',
        label: 'Close Workspace'
    };
    WorkspaceCommands.NEW_FILE = {
        id: 'file.newFile',
        label: 'New File'
    };
    WorkspaceCommands.NEW_FOLDER = {
        id: 'file.newFolder',
        label: 'New Folder'
    };
    WorkspaceCommands.FILE_OPEN_WITH = function (opener) { return ({
        id: "file.openWith." + opener.id,
        label: opener.label,
        iconClass: opener.iconClass
    }); };
    WorkspaceCommands.FILE_RENAME = {
        id: 'file.rename',
        label: 'Rename'
    };
    WorkspaceCommands.FILE_DELETE = {
        id: 'file.delete',
        label: 'Delete'
    };
    WorkspaceCommands.FILE_COMPARE = {
        id: 'file.compare',
        label: 'Compare with Each Other'
    };
})(WorkspaceCommands = exports.WorkspaceCommands || (exports.WorkspaceCommands = {}));
var FileMenuContribution = /** @class */ (function () {
    function FileMenuContribution() {
    }
    FileMenuContribution.prototype.registerMenus = function (registry) {
        registry.registerMenuAction(common_frontend_contribution_1.CommonMenus.FILE_NEW, {
            commandId: WorkspaceCommands.NEW_FILE.id
        });
        registry.registerMenuAction(common_frontend_contribution_1.CommonMenus.FILE_NEW, {
            commandId: WorkspaceCommands.NEW_FOLDER.id
        });
    };
    FileMenuContribution = __decorate([
        inversify_1.injectable()
    ], FileMenuContribution);
    return FileMenuContribution;
}());
exports.FileMenuContribution = FileMenuContribution;
var WorkspaceCommandContribution = /** @class */ (function () {
    function WorkspaceCommandContribution(fileSystem, workspaceService, selectionService, openerService, app, messageService) {
        this.fileSystem = fileSystem;
        this.workspaceService = workspaceService;
        this.selectionService = selectionService;
        this.openerService = openerService;
        this.app = app;
        this.messageService = messageService;
    }
    WorkspaceCommandContribution.prototype.registerCommands = function (registry) {
        var _this = this;
        this.openerService.getOpeners().then(function (openers) {
            var e_1, _a;
            var _loop_1 = function (opener_1) {
                var openWithCommand = WorkspaceCommands.FILE_OPEN_WITH(opener_1);
                registry.registerCommand(openWithCommand, _this.newUriAwareCommandHandler({
                    execute: function (uri) { return opener_1.open(uri); },
                    isEnabled: function (uri) { return opener_1.canHandle(uri) > 0; },
                    isVisible: function (uri) { return opener_1.canHandle(uri) > 0; }
                }));
            };
            try {
                for (var openers_1 = __values(openers), openers_1_1 = openers_1.next(); !openers_1_1.done; openers_1_1 = openers_1.next()) {
                    var opener_1 = openers_1_1.value;
                    _loop_1(opener_1);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (openers_1_1 && !openers_1_1.done && (_a = openers_1.return)) _a.call(openers_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
        registry.registerCommand(WorkspaceCommands.NEW_FILE, this.newWorkspaceRootUriAwareCommandHandler({
            execute: function (uri) { return _this.getDirectory(uri).then(function (parent) {
                if (parent) {
                    var parentUri_1 = new uri_1.default(parent.uri);
                    var vacantChildUri = _this.findVacantChildUri(parentUri_1, parent, 'Untitled', '.txt');
                    var dialog = new dialogs_1.SingleTextInputDialog({
                        title: "New File",
                        initialValue: vacantChildUri.path.base,
                        validate: function (name) { return _this.validateFileName(name, parent); }
                    });
                    dialog.open().then(function (name) {
                        var fileUri = parentUri_1.resolve(name);
                        _this.fileSystem.createFile(fileUri.toString()).then(function () {
                            browser_1.open(_this.openerService, fileUri);
                        });
                    });
                }
            }); }
        }));
        registry.registerCommand(WorkspaceCommands.NEW_FOLDER, this.newWorkspaceRootUriAwareCommandHandler({
            execute: function (uri) { return _this.getDirectory(uri).then(function (parent) {
                if (parent) {
                    var parentUri_2 = new uri_1.default(parent.uri);
                    var vacantChildUri = _this.findVacantChildUri(parentUri_2, parent, 'Untitled');
                    var dialog = new dialogs_1.SingleTextInputDialog({
                        title: "New Folder",
                        initialValue: vacantChildUri.path.base,
                        validate: function (name) { return _this.validateFileName(name, parent); }
                    });
                    dialog.open().then(function (name) {
                        return _this.fileSystem.createFolder(parentUri_2.resolve(name).toString());
                    });
                }
            }); }
        }));
        registry.registerCommand(WorkspaceCommands.FILE_RENAME, this.newUriAwareCommandHandler({
            execute: function (uri) { return _this.getParent(uri).then(function (parent) {
                if (parent) {
                    var dialog = new dialogs_1.SingleTextInputDialog({
                        title: 'Rename File',
                        initialValue: uri.path.base,
                        validate: function (name) { return _this.validateFileName(name, parent); }
                    });
                    dialog.open().then(function (name) {
                        return _this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString());
                    });
                }
            }); }
        }));
        var rootUri;
        this.workspaceService.root.then(function (root) {
            if (root) {
                rootUri = new uri_1.default(root.uri);
            }
        });
        registry.registerCommand(WorkspaceCommands.FILE_DELETE, this.newMultiUriAwareCommandHandler({
            isVisible: function (uris) { return !(rootUri && uris.some(function (uri) { return uri.toString() === rootUri.toString(); })); },
            execute: function (uris) { return __awaiter(_this, void 0, void 0, function () {
                var msg, dialog;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            msg = (function () {
                                var e_2, _a;
                                if (uris.length === 1) {
                                    return "Do you really want to delete " + uris[0].path.base + "?";
                                }
                                if (uris.length > 10) {
                                    return "Do you really want to delete all the " + uris.length + " selected files?";
                                }
                                var messageContainer = document.createElement('div');
                                messageContainer.textContent = 'Do you really want to delete the following files?';
                                var list = document.createElement('ul');
                                list.style.listStyleType = 'none';
                                try {
                                    for (var uris_1 = __values(uris), uris_1_1 = uris_1.next(); !uris_1_1.done; uris_1_1 = uris_1.next()) {
                                        var uri = uris_1_1.value;
                                        var listItem = document.createElement('li');
                                        listItem.textContent = uri.path.base;
                                        list.appendChild(listItem);
                                    }
                                }
                                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                                finally {
                                    try {
                                        if (uris_1_1 && !uris_1_1.done && (_a = uris_1.return)) _a.call(uris_1);
                                    }
                                    finally { if (e_2) throw e_2.error; }
                                }
                                messageContainer.appendChild(list);
                                return messageContainer;
                            })();
                            dialog = new dialogs_1.ConfirmDialog({
                                title: "Delete File" + (uris.length === 1 ? '' : 's'),
                                msg: msg,
                            });
                            return [4 /*yield*/, dialog.open()];
                        case 1:
                            if (!_a.sent()) return [3 /*break*/, 3];
                            // Make sure we delete the longest paths first, they might be nested. Longer paths come first.
                            uris.sort(function (left, right) { return right.toString().length - left.toString().length; });
                            return [4 /*yield*/, Promise.all(uris.map(function (uri) { return uri.toString(); }).map(function (uri) { return _this.fileSystem.delete(uri); }))];
                        case 2:
                            _a.sent();
                            _a.label = 3;
                        case 3: return [2 /*return*/];
                    }
                });
            }); }
        }));
        registry.registerCommand(WorkspaceCommands.FILE_COMPARE, this.newMultiUriAwareCommandHandler({
            isVisible: function (uris) { return uris.length === 2; },
            execute: function (uris) { return __awaiter(_this, void 0, void 0, function () {
                var _a, left, right, _b, leftExists, rightExists, _c, leftStat_1, rightStat_1, uri, opener_2, details;
                return __generator(this, function (_d) {
                    switch (_d.label) {
                        case 0:
                            _a = __read(uris, 2), left = _a[0], right = _a[1];
                            return [4 /*yield*/, Promise.all([
                                    this.fileSystem.exists(left.toString()),
                                    this.fileSystem.exists(right.toString())
                                ])];
                        case 1:
                            _b = __read.apply(void 0, [_d.sent(), 2]), leftExists = _b[0], rightExists = _b[1];
                            if (!(leftExists && rightExists)) return [3 /*break*/, 5];
                            return [4 /*yield*/, Promise.all([
                                    this.fileSystem.getFileStat(left.toString()),
                                    this.fileSystem.getFileStat(right.toString()),
                                ])];
                        case 2:
                            _c = __read.apply(void 0, [_d.sent(), 2]), leftStat_1 = _c[0], rightStat_1 = _c[1];
                            if (!(leftStat_1 && rightStat_1)) return [3 /*break*/, 5];
                            if (!(!leftStat_1.isDirectory && !rightStat_1.isDirectory)) return [3 /*break*/, 4];
                            uri = diff_uris_1.DiffUris.encode(left, right);
                            return [4 /*yield*/, this.openerService.getOpener(uri)];
                        case 3:
                            opener_2 = _d.sent();
                            opener_2.open(uri);
                            return [3 /*break*/, 5];
                        case 4:
                            details = (function () {
                                if (leftStat_1.isDirectory && rightStat_1.isDirectory) {
                                    return 'Both resource were a directory.';
                                }
                                else {
                                    if (leftStat_1.isDirectory) {
                                        return "'" + left.path.base + "' was a directory.";
                                    }
                                    else {
                                        return "'" + right.path.base + "' was a directory.";
                                    }
                                }
                            });
                            this.messageService.warn("Directories cannot be compared. " + details());
                            _d.label = 5;
                        case 5: return [2 /*return*/];
                    }
                });
            }); }
        }));
    };
    WorkspaceCommandContribution.prototype.newUriAwareCommandHandler = function (handler) {
        return new uri_command_handler_1.UriAwareCommandHandler(this.selectionService, handler);
    };
    WorkspaceCommandContribution.prototype.newMultiUriAwareCommandHandler = function (handler) {
        return new uri_command_handler_1.UriAwareCommandHandler(this.selectionService, handler, { multi: true });
    };
    WorkspaceCommandContribution.prototype.newWorkspaceRootUriAwareCommandHandler = function (handler) {
        return new WorkspaceRootUriAwareCommandHandler(this.workspaceService, this.selectionService, handler);
    };
    /**
     * Returns an error message if the file name is invalid. Otherwise, an empty string.
     *
     * @param name the simple file name of the file to validate.
     * @param parent the parent directory's file stat.
     */
    WorkspaceCommandContribution.prototype.validateFileName = function (name, parent) {
        var e_3, _a;
        if (!validFilename(name)) {
            return "Invalid name, try other";
        }
        if (parent.children) {
            try {
                for (var _b = __values(parent.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var child = _c.value;
                    if (new uri_1.default(child.uri).path.base === name) {
                        return 'A file with this name already exists.';
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        return '';
    };
    WorkspaceCommandContribution.prototype.getDirectory = function (candidate) {
        return __awaiter(this, void 0, void 0, function () {
            var stat;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.fileSystem.getFileStat(candidate.toString())];
                    case 1:
                        stat = _a.sent();
                        if (stat && stat.isDirectory) {
                            return [2 /*return*/, stat];
                        }
                        return [2 /*return*/, this.getParent(candidate)];
                }
            });
        });
    };
    WorkspaceCommandContribution.prototype.getParent = function (candidate) {
        return this.fileSystem.getFileStat(candidate.parent.toString());
    };
    WorkspaceCommandContribution.prototype.findVacantChildUri = function (parentUri, parent, name, ext) {
        if (ext === void 0) { ext = ''; }
        var children = !parent.children ? [] : parent.children.map(function (child) { return new uri_1.default(child.uri); });
        var index = 1;
        var base = name + ext;
        while (children.some(function (child) { return child.path.base === base; })) {
            index = index + 1;
            base = name + '_' + index + ext;
        }
        return parentUri.resolve(base);
    };
    WorkspaceCommandContribution = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(filesystem_1.FileSystem)),
        __param(1, inversify_1.inject(workspace_service_1.WorkspaceService)),
        __param(2, inversify_1.inject(selection_service_1.SelectionService)),
        __param(3, inversify_1.inject(browser_1.OpenerService)),
        __param(4, inversify_1.inject(browser_1.FrontendApplication)),
        __param(5, inversify_1.inject(message_service_1.MessageService)),
        __metadata("design:paramtypes", [Object, workspace_service_1.WorkspaceService,
            selection_service_1.SelectionService, Object, browser_1.FrontendApplication,
            message_service_1.MessageService])
    ], WorkspaceCommandContribution);
    return WorkspaceCommandContribution;
}());
exports.WorkspaceCommandContribution = WorkspaceCommandContribution;
var WorkspaceRootUriAwareCommandHandler = /** @class */ (function (_super) {
    __extends(WorkspaceRootUriAwareCommandHandler, _super);
    function WorkspaceRootUriAwareCommandHandler(workspaceService, selectionService, handler) {
        var _this = _super.call(this, selectionService, handler) || this;
        _this.workspaceService = workspaceService;
        _this.selectionService = selectionService;
        _this.handler = handler;
        workspaceService.root.then(function (root) {
            if (root) {
                _this.rootUri = new uri_1.default(root.uri);
            }
        });
        return _this;
    }
    WorkspaceRootUriAwareCommandHandler.prototype.getUri = function () {
        return _super.prototype.getUri.call(this) || this.rootUri;
    };
    return WorkspaceRootUriAwareCommandHandler;
}(uri_command_handler_1.UriAwareCommandHandler));
exports.WorkspaceRootUriAwareCommandHandler = WorkspaceRootUriAwareCommandHandler;
