// @theia/languages/lib/browser/language-client-factory.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var browser_1 = require("@theia/core/lib/browser");
var base_1 = require("vscode-base-languageclient/lib/base");
var common_1 = require("../common");
var LanguageClientFactory = /** @class */ (function () {
    function LanguageClientFactory(workspace, languages, window, connectionProvider) {
        this.workspace = workspace;
        this.languages = languages;
        this.window = window;
        this.connectionProvider = connectionProvider;
    }
    LanguageClientFactory.prototype.get = function (contribution, clientOptions) {
        var _this = this;
        var _a = this, workspace = _a.workspace, languages = _a.languages, window = _a.window;
        var commands = clientOptions.commands;
        var services = { workspace: workspace, languages: languages, commands: commands, window: window };
        if (!clientOptions.errorHandler) {
            clientOptions.errorHandler = {
                // ignore connection errors
                error: function () { return base_1.ErrorAction.Continue; },
                closed: function () { return defaultErrorHandler.closed(); }
            };
        }
        var client = new common_1.BaseLanguageClient({
            name: contribution.name,
            clientOptions: clientOptions,
            services: services,
            connectionProvider: {
                get: function (errorHandler, closeHandler) {
                    return new Promise(function (resolve) {
                        _this.connectionProvider.listen({
                            path: common_1.LanguageContribution.getPath(contribution),
                            onConnection: function (messageConnection) {
                                var connection = common_1.createConnection(messageConnection, errorHandler, closeHandler);
                                resolve(connection);
                            },
                        }, { reconnecting: false });
                    });
                }
            }
        });
        var defaultErrorHandler = client.createDefaultErrorHandler();
        return client;
    };
    LanguageClientFactory = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(common_1.Workspace)),
        __param(1, inversify_1.inject(common_1.Languages)),
        __param(2, inversify_1.inject(common_1.Window)),
        __param(3, inversify_1.inject(browser_1.WebSocketConnectionProvider)),
        __metadata("design:paramtypes", [Object, Object, Object, browser_1.WebSocketConnectionProvider])
    ], LanguageClientFactory);
    return LanguageClientFactory;
}());
exports.LanguageClientFactory = LanguageClientFactory;
