// @theia/core/lib/common/messaging/web-socket-channel.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
var disposable_1 = require("../disposable");
var WebSocketChannel = /** @class */ (function () {
    function WebSocketChannel(id, doSend) {
        this.id = id;
        this.doSend = doSend;
        this.toDispose = new disposable_1.DisposableCollection();
        this.fireOpen = function () { };
        this.fireMessage = function () { };
        this.fireError = function () { };
        this.fireClose = function () { };
        this.toDispose.push(disposable_1.Disposable.NULL);
    }
    WebSocketChannel.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    WebSocketChannel.prototype.checkNotDisposed = function () {
        if (this.toDispose.disposed) {
            throw new Error('The channel has been disposed.');
        }
    };
    WebSocketChannel.prototype.handleMessage = function (message) {
        if (message.kind === 'ready') {
            this.fireOpen();
        }
        else if (message.kind === 'data') {
            this.fireMessage(message.content);
        }
        else if (message.kind === 'close') {
            this.fireClose(1000, '');
        }
    };
    WebSocketChannel.prototype.open = function (path) {
        this.checkNotDisposed();
        this.doSend(JSON.stringify({
            kind: 'open',
            id: this.id,
            path: path
        }));
    };
    WebSocketChannel.prototype.ready = function () {
        this.checkNotDisposed();
        this.doSend(JSON.stringify({
            kind: 'ready',
            id: this.id
        }));
    };
    WebSocketChannel.prototype.send = function (content) {
        this.checkNotDisposed();
        this.doSend(JSON.stringify({
            kind: 'data',
            id: this.id,
            content: content
        }));
    };
    WebSocketChannel.prototype.close = function () {
        this.checkNotDisposed();
        this.doSend(JSON.stringify({
            kind: 'close',
            id: this.id
        }));
    };
    WebSocketChannel.prototype.onOpen = function (cb) {
        var _this = this;
        this.checkNotDisposed();
        this.fireOpen = cb;
        this.toDispose.push(disposable_1.Disposable.create(function () { return _this.fireOpen = function () { }; }));
    };
    WebSocketChannel.prototype.onMessage = function (cb) {
        var _this = this;
        this.checkNotDisposed();
        this.fireMessage = cb;
        this.toDispose.push(disposable_1.Disposable.create(function () { return _this.fireMessage = function () { }; }));
    };
    WebSocketChannel.prototype.onError = function (cb) {
        var _this = this;
        this.checkNotDisposed();
        this.fireError = cb;
        this.toDispose.push(disposable_1.Disposable.create(function () { return _this.fireError = function () { }; }));
    };
    WebSocketChannel.prototype.onClose = function (cb) {
        var _this = this;
        this.checkNotDisposed();
        this.fireClose = cb;
        this.toDispose.push(disposable_1.Disposable.create(function () { return _this.fireClose = function () { }; }));
    };
    WebSocketChannel.wsPath = '/services';
    return WebSocketChannel;
}());
exports.WebSocketChannel = WebSocketChannel;
